packages/@react-stately/table/src/useTreeGridState.ts

```diff
@@ -0,0 +1,305 @@
+/*
+ * Copyright 2023 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {CollectionBuilder} from '@react-stately/collections';
+import {Expandable} from '@react-types/shared';
+import {GridNode} from '@react-types/grid';
+import {Key, useMemo} from 'react';
+import {TableCollection} from './TableCollection';
+import {TableState, TableStateProps, useTableState} from './useTableState';
+import {useControlledState} from '@react-stately/utils';
+
+export interface TreeGridState<T> extends TableState<T> {
+  /** A set of keys for items that are expanded. */
+  expandedKeys: 'all' | Set<Key>,
+  /** Toggles the expanded state for a row by its key. */
+  toggleKey(key: Key): void,
+  /** The key map containing nodes representing the collection's tree grid structure. */
+  keyMap: Map<Key, GridNode<T>>,
+  /** The number of columns provided by the user. */
+  userColumnCount: number
+}
+
+export interface TreeGridStateProps<T> extends Expandable, Omit<TableStateProps<T>, 'collection'> {}
+
+/**
+ * Provides state management for a tree grid component. Handles building a collection
+ * of columns and rows from props. In addition, it tracks and manages expanded rows, row selection, and sort order changes.
+ */
+export function useTreeGridState<T extends object>(props: TreeGridStateProps<T>): TreeGridState<T> {
```

Comment:
prefix this with UNSTABLE and add an error throw if feature flag isn't turned on